test/threadstest.c

```diff
@@ -741,6 +745,309 @@ static int test_bio_dgram_pair(void)
+        return 0;
+    if (!TEST_int_eq(ossl_crypto_thread_native_clean(t), 1))
+        return 0;
+
+    ossl_crypto_mutex_unlock(lock);
+    ossl_crypto_mutex_free(&lock);
+    if (!TEST_ptr_null(lock))
+        return 0;
+
+    return 1;
+}
+
+#if !defined(OPENSSL_NO_DEFAULT_THREAD_POOL)
+static int test_thread_internal(void)
+{
+    uint32_t retval[3];
+    uint32_t local[3] = { 0 };
+    uint32_t threads_supported;
+    size_t i;
+    void *t[3];
+    OSSL_LIB_CTX *cust_ctx = OSSL_LIB_CTX_new();
+
+    threads_supported = OSSL_get_thread_support_flags();
+    threads_supported &= OSSL_THREAD_SUPPORT_FLAG_DEFAULT_SPAWN;
+
+    /* currently only the default pool is supported */
+    if (threads_supported == 0) {
+        if (!TEST_uint64_t_eq(OSSL_get_max_threads(NULL), 0))
+            return 0;
+        if (!TEST_uint64_t_eq(OSSL_get_max_threads(cust_ctx), 0))
+            return 0;
+
+        if (!TEST_int_eq(OSSL_set_max_threads(NULL, 1), 0))
+            return 0;
+        if (!TEST_int_eq(OSSL_set_max_threads(cust_ctx, 1), 0))
+            return 0;
+
+        if (!TEST_uint64_t_eq(OSSL_get_max_threads(NULL), 0))
+            return 0;
+        if (!TEST_uint64_t_eq(OSSL_get_max_threads(cust_ctx), 0))
+            return 0;
+
+        t[0] = ossl_crypto_thread_start(NULL, test_thread_native_fn, &local[0]);
+        if (!TEST_ptr_null(t[0]))
+            return 0;
+        
+        return 1;
+    }
+
+    /* fail when not allowed to use threads */
+
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(NULL), 0))
+        return 0;
+    t[0] = ossl_crypto_thread_start(NULL, test_thread_native_fn, &local[0]);
+    if (!TEST_ptr_null(t[0]))
+        return 0;
+
+    /* fail when enabled on a different context */
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(cust_ctx), 0))
+        return 0;
+    if (!TEST_int_eq(OSSL_set_max_threads(cust_ctx, 1), 1))
+        return 0;
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(NULL), 0))
+        return 0;
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(cust_ctx), 1))
+        return 0;
+    t[0] = ossl_crypto_thread_start(NULL, test_thread_native_fn, &local[0]);
+    if (!TEST_ptr_null(t[0]))
+        return 0;
+    if (!TEST_int_eq(OSSL_set_max_threads(cust_ctx, 0), 1))
+        return 0;
+
+    /* sequential startup */
+
+    if (!TEST_int_eq(OSSL_set_max_threads(NULL, 1), 1))
+        return 0;
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(NULL), 1))
+        return 0;
+    if (!TEST_uint64_t_eq(OSSL_get_max_threads(cust_ctx), 0))
+        return 0;
+
+    for (i = 0; i < OSSL_NELEM(t); ++i) {
+        local[0] = i + 1;
+
+        t[i] = ossl_crypto_thread_start(NULL, test_thread_native_fn, &local[0]);
+        if (!TEST_ptr(t[i]))
+            return 0;
+
+        if (!TEST_int_eq(ossl_crypto_thread_join(t[i], &retval[0]), 1))
+            return 0;
+        if (!TEST_int_eq(ossl_crypto_thread_join(t[i], &retval[0]), 1))
+            return 0;
```

Comment:
```
pthread_join(3): Joining with a thread that has previously been joined results in undefined behavior.
```
Since we do not restrict the user from doing so, I thought it best to test it; threadsanitizer would jump at us if we did not handle it properly and did not respect pthread api:
```c
#include <pthread.h>
void *foo(void *p) {
  pthread_exit(p);
}
int main(void) {
  pthread_t id;
  pthread_create(&id, NULL, foo, NULL);
  pthread_join(id, NULL);
#ifdef DUP
  pthread_join(id, NULL);
#endif
}
```
```sh
# clang -Wall -Werror -pedantic -fsanitize=thread ./test.c -o test ; ./test
# clang -Wall -Werror -pedantic -fsanitize=thread ./test.c -o test -DDUP ; ./test
ThreadSanitizer: CHECK failed: sanitizer_thread_registry.cpp:348 "((t)) != (0)" (0x0, 0x0) (tid=256762)
    #0 __tsan::CheckUnwind() <null> (test+0x4bd3c4) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
    #1 __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) <null> (test+0x434cd4) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
    #2 __sanitizer::ThreadRegistry::ConsumeThreadUserId(unsigned long) <null> (test+0x43346c) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
    #3 pthread_join <null> (test+0x4660fc) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
    #4 main <null> (test+0x4e076c) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
    #5 __libc_start_call_main <null> (libc.so.6+0x273c4) (BuildId: e4d169ff5a558b428c64c9dbfdc6c02f61d4f4ff)
    #6 __libc_start_main@GLIBC_2.17 <null> (libc.so.6+0x2749c) (BuildId: e4d169ff5a558b428c64c9dbfdc6c02f61d4f4ff)
    #7 _start <null> (test+0x41ef6c) (BuildId: 62ae8af4a7cea577224141a033ed9561f3d0ee61)
```

Reply:
Perhaps add a comment saying who you are doing this?